bbot/modules/internal/excavate.py

```diff
@@ -1,476 +1,858 @@
+
+    params = parse_qs(parsed_url.query)
+    flat_params = {k: v[0] for k, v in params.items()}
+
+    for p, p_value in flat_params.items():
+        yield "GET", parsed_url, p, p_value, "location_header", _exclude_key(flat_params, p)
+
+
+class YaraRuleSettings:
+
+    def __init__(self, description, tags, emit_match):
+        self.description = description
+        self.tags = tags
+        self.emit_match = emit_match
+
+
+class ExcavateRule:
+    """
+    The BBOT Regex Commandments:
 
+    1) Thou shalt employ YARA regexes in place of Python regexes, save when necessity doth compel otherwise.
+    2) Thou shalt ne'er wield a Python regex against a vast expanse of text.
+    3) Whensoever it be possible, thou shalt favor string matching o'er regexes.
 
-class BaseExtractor:
-    # If using capture groups, be sure to name them beginning with "capture".
-    regexes = {}
+    Amen.
+    """
+
+    yara_rules = {}
 
     def __init__(self, excavate):
         self.excavate = excavate
         self.helpers = excavate.helpers
-        self.compiled_regexes = {}
-        for rname, r in self.regexes.items():
-            self.compiled_regexes[rname] = re.compile(r)
-
-    async def search(self, content, event, **kwargs):
-        results = set()
-        async for result, name in self._search(content, event, **kwargs):
-            results.add((result, name))
-        for result, name in results:
-            await self.report(result, name, event, **kwargs)
-
-    async def _search(self, content, event, **kwargs):
-        for name, regex in self.compiled_regexes.items():
-            # yield to event loop
-            await self.excavate.helpers.sleep(0)
-            for result in await self.helpers.re.findall(regex, content):
-                yield result, name
-
-    async def report(self, result, name, event):
-        pass
-
-
-class CSPExtractor(BaseExtractor):
-    regexes = {"CSP": r"(?i)(?m)Content-Security-Policy:.+$"}
-
-    async def extract_domains(self, csp):
-        domains = await self.helpers.re.findall(dns_name_regex, csp)
-        unique_domains = set(domains)
-        return unique_domains
-
-    async def search(self, content, event, **kwargs):
-        async for csp, name in self._search(content, event, **kwargs):
-            extracted_domains = await self.extract_domains(csp)
-            for domain in extracted_domains:
-                await self.report(domain, event, **kwargs)
-
-    async def report(self, domain, event, **kwargs):
-        await self.excavate.emit_event(
-            domain,
-            "DNS_NAME",
-            parent=event,
-            tags=["affiliate"],
-            context=f"{{module}}'s CSP extractor found {{event.type}}: {{event.data}} in CSP header",
-        )
-
-
-class HostnameExtractor(BaseExtractor):
-    regexes = {}
+        self.name = ""
+
+    async def preprocess(self, r, event, discovery_context):
+        self.discovery_context = discovery_context
```

Comment:
Since the same rule might be used to process multiple events concurrently, we need to avoid setting event-specific data on `self`.

Reply:
https://github.com/blacklanternsecurity/bbot/pull/1471/commits/b191770fcb157ac1b7043881ba0b83edf77b56b2